codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/DafnyToSmithyShapeVisitor.java

```diff
@@ -349,17 +350,59 @@ public String doubleShape(DoubleShape shape) {
     @Override
     public String unionShape(UnionShape shape) {
         writer.addImportFromModule("github.com/dafny-lang/DafnyRuntimeGo", "dafny");
-        return """
-            func () types.%s {
-                var union types.%s
-                union = &types.MyUnionMemberIntegerValue{
-                    Value: %s.(int32),
-                }
-                return union
-            }()""".formatted(
-                shape.toShapeId().getName(),
-                shape.toShapeId().getName(),
-                dataSource);
+
+        String returnString = """
+                func() %s {
+                """.formatted(context.symbolProvider().toSymbol(shape));
+        returnString += """
+                    if %s == nil {
```

Comment:
Are unions always Optional? Otherwise this nil check will fail.

Reply:
I made a fix for it.